### PR TITLE
Support multiple metadata transport protocols

### PR DESCRIPTION
**Pull request by masih**
Introduce the `Transport` type to capture information about multiple
transport protocols within a single metadata.

Re implement `Metadata` type to accommodate:
* multiple transport protocols
* enforce they are sorted by their ID
* backward compatibility with existing metadata format
* deserialization of existing formats gracefully.

~Add metadata conversion to keep code compiling until this PR is merged:~
 ~- https://github.com/filecoin-project/storetheindex/pull/250~

The PR now depends on the head of sti `main` now that [this](https://github.com/filecoin-project/storetheindex/pull/260) PR is merged.

Fixes: #188

### CODE REVIEW

**Comment by masih** [comment]
@gammazero  I think this PR would make #196 redundant if it only contains multi metadata support

File: metadata/graphsync_filecoinv1.go [hunk 1]
```diff
@@ -3,17 +3,25 @@ package metadata
...
-type GraphsyncFilecoinV1Metadata struct {
+var (
+	_ Transport = (*GraphsyncFilecoinV1Transport)(nil)
+
+	graphSyncfilecoinV1SchemaType schema.Type
+	graphSyncfilecoinV1Prototype  schema.TypedPrototype
+)
+
+// GraphsyncFilecoinV1Transport represents the indexing metadata for multicodec.TransportGraphsyncFilecoinv1.
+type GraphsyncFilecoinV1Transport struct {
```
[the remark this answers is not in the document]
Comment (masih) [reply metadata/graphsync_filecoinv1.go]: Dropped suffixes alltogether in favour of `metadata.Bitswap` and `metata.GraphsyncFilecoinV1`. See what you think.

[the remark this answers is not in the document]
Comment (gammazero) [reply metadata/graphsync_filecoinv1.go]: I like that best.

**Comment by masih** [comment]
Marking as draft until this is merged: https://github.com/filecoin-project/storetheindex/pull/260

File: metadata/metadata.go [hunk 2]
```diff
@@ -0,0 +1,154 @@
...
+// Equals checks whether this Metadata is equal with the other Metadata.
+// The two are considered equal if they have the same transports with the same order.
+func (m *Metadata) Equals(other *Metadata) bool {
+	if len(m.protocols) != len(other.protocols) {
+		return false
+	}
+	for i, this := range m.protocols {
+		that := other.protocols[i]
+		// TODO: Is it more efficient to check fields individually?
+		if !reflect.DeepEqual(this, that) {
```
Comment (gammazero) [inline metadata/metadata.go:138]: This is probably less efficient as it relies on Reflect.  The check should call to the `Equals` method the `Protocol`.

**Review by gammazero — APPROVED** [review]
Awesome! See comments.

File: cmd/provider/flags.go [hunk 3]
```diff
@@ -73,17 +73,7 @@ var indexFlags = []cli.Flag{
...
-		Usage:    "Metadata bytes.",
-		Aliases:  []string{"m"},
-		Required: false,
-	},
-	&cli.IntFlag{
-		Name:     "proto",
-		Usage:    "Specify retrieval protocol ID",
-		Required: true,
-	},
+	metadataFlag,
```
[the remark this answers is not in the document]
Comment (masih) [reply cmd/provider/flags.go:76]: Captured: https://github.com/filecoin-project/index-provider/issues/199

**Pull request merged**